syntax = "proto2";
import "google/protobuf/descriptor.proto";

package goby;

extend .google.protobuf.FieldOptions
{
    // Email from liujisi@google.com on 7.22.2011: Please use 1009 for all types
    // of options.
    // --------------------------------
    // Project:  Goby Underwater Autonomy Project
    // Contact:  Toby Schneider <tes@mit.edu>
    // Web site: https://launchpad.net/goby
    // Extensions: 1009 (all types)
    // --------------------------------
    optional GobyFieldOptions field = 1009;
}

extend .google.protobuf.MessageOptions
{
    optional GobyMessageOptions msg = 1009;
}

extend .google.protobuf.EnumValueOptions
{
    optional GobyEnumValueOptions ev = 1009;
}

message GobyFieldOptions
{
    optional string description = 100;
    optional string example = 101;

    // used in goby_moos_app.proto
    optional string moos_global = 102;

    message ConfigurationOptions
    {
        optional string cli_short = 1;
        enum ConfigAction
        {
            ALWAYS = 1;
            ADVANCED = 2;
            DEVELOPER = 3;
            HIDDEN = 4;
            NEVER = 5;
        }
        optional ConfigAction action = 2 [default = ALWAYS];
        optional int32 position = 3;
    }
    optional ConfigurationOptions cfg = 200;
}

message GobyMessageOptions
{
    optional string convertible_from =
        100;  // tags this message as convertible from another existing message
              // type based on field names

    message ConfigurationOptions
    {
        optional bool tool_mode = 1
            [default =
                 false];  // if true, reads only parameters up to the first
                          // positional parameter (assumed to be an action)
        optional bool has_help_action = 2
            [default = false];  // if true, do not output "Usage"  with the default --help, and do not exit after auto help
    }
    optional ConfigurationOptions cfg = 200;
}

message GobyEnumValueOptions
{
    message ConfigurationOptions
    {
        optional string external_command = 1;
        optional string external_help_param = 2 [default = "-h"];
        optional string short_help_msg = 3;
        optional string full_help_msg = 4;
    }
    optional ConfigurationOptions cfg = 200;
}
